src/main.cpp

```diff
@@ -5088,55 +5105,62 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
-                            // Note that there is currently no way for a node to request any single transactions we didn't send here -
-                            // they must either disconnect and retry or request the full block.
-                            // Thus, the protocol spec specified allows for us to provide duplicate txn here,
-                            // however we MUST always provide at least what the remote peer needs
-                            typedef std::pair<unsigned int, uint256> PairType;
-                            BOOST_FOREACH(PairType& pair, merkleBlock.vMatchedTxn)
-			      {
-       	                        pfrom->txsSent += 1;
-                                pfrom->PushMessage(NetMsgType::TX, block.vtx[pair.first]);
-			      }
+                            pfrom->PushMessage(NetMsgType::BLOCK, block);
                         }
-                        // else
+
+                        // BUIP010 Xtreme Thinblocks: begin section
+                        else if (inv.type == MSG_THINBLOCK || inv.type == MSG_XTHINBLOCK) {
+                          LogPrint("thin", "Sending xthin by INV queue getdata message\n");
+                          SendXThinBlock(block, pfrom, inv);
+                        }
+                        // BUIP010 Xtreme Thinblocks: end section
+
+                        else // MSG_FILTERED_BLOCK)
+                        {
+                            LOCK(pfrom->cs_filter);
+                            if (pfrom->pfilter)
+                            {
+                                CMerkleBlock merkleBlock(block, *pfrom->pfilter);
+                                pfrom->PushMessage(NetMsgType::MERKLEBLOCK, merkleBlock);
+                                pfrom->blocksSent += 1;
+                                // CMerkleBlock just contains hashes, so also push any transactions in the block the client did not see
+                                // This avoids hurting performance by pointlessly requiring a round-trip
+                                // Note that there is currently no way for a node to request any single transactions we didn't send here -
+                                // they must either disconnect and retry or request the full block.
+                                // Thus, the protocol spec specified allows for us to provide duplicate txn here,
+                                // however we MUST always provide at least what the remote peer needs
+                                typedef std::pair<unsigned int, uint256> PairType;
+                                BOOST_FOREACH(PairType& pair, merkleBlock.vMatchedTxn)
+                                {
+                                    pfrom->txsSent += 1;
+                                    pfrom->PushMessage(NetMsgType::TX, block.vtx[pair.first]);
+                                }
+                            }
+                            // else
```

Comment:
It looks like looking from the diff that this whole code was rewritten, but it ain't thee case. This makes it impossible to spot what actually changed. please split actual code change and moving code around in different PR. this is unreviewable in this form.